File: expressjs/src/storage/local.ts
import fs from "node:fs";
import fsp from "node:fs/promises";
import path from "node:path";
import type { FileStorage } from "./storage.js";

/** Local filesystem storage implementation. */
export class LocalStorage implements FileStorage {
  constructor(private basePath: string) {}

  async save(appName: string, fileID: string, filename: string, buffer: Buffer): Promise<string> {
    const dir = path.join(this.basePath, appName, fileID);
    await fsp.mkdir(dir, { recursive: true });

    const storagePath = path.join(dir, filename);
    await fsp.writeFile(storagePath, buffer);
    return storagePath;
  }

  async open(storagePath: string): Promise<Buffer> {
    return fsp.readFile(storagePath);
  }

  async delete(storagePath: string): Promise<void> {
    try {
      await fsp.unlink(storagePath);
    } catch (err: any) {
      if (err.code !== "ENOENT") throw err;
    }
    // Try to remove parent dir (fileID dir) if empty
    const dir = path.dirname(storagePath);
    try {
      await fsp.rmdir(dir);
    } catch {
      // ignore — dir not empty or already gone
    }
  }
}

File: golang/internal/storage/local.go
package storage

import (
	"context"
	"fmt"
	"io"
	"os"
	"path/filepath"
)

// LocalStorage stores files on the local filesystem.
type LocalStorage struct {
	basePath string
}

func NewLocalStorage(basePath string) *LocalStorage {
	return &LocalStorage{basePath: basePath}
}

func (s *LocalStorage) Save(_ context.Context, appName, fileID, filename string, reader io.Reader) (string, error) {
	dir := filepath.Join(s.basePath, appName, fileID)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return "", fmt.Errorf("create dir: %w", err)
	}

	storagePath := filepath.Join(dir, filename)
	f, err := os.Create(storagePath)
	if err != nil {
		return "", fmt.Errorf("create file: %w", err)
	}
	defer f.Close()

	if _, err := io.Copy(f, reader); err != nil {
		return "", fmt.Errorf("write file: %w", err)
	}

	return storagePath, nil
}

func (s *LocalStorage) Open(_ context.Context, storagePath string) (io.ReadCloser, error) {
	f, err := os.Open(storagePath)
	if err != nil {
		return nil, fmt.Errorf("open file: %w", err)
	}
	return f, nil
}

func (s *LocalStorage) Delete(_ context.Context, storagePath string) error {
	if err := os.Remove(storagePath); err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("remove file: %w", err)
	}
	// Try to remove parent dir (fileID dir) if empty
	dir := filepath.Dir(storagePath)
	_ = os.Remove(dir)
	return nil
}
